main.tf

```diff
@@ -15,99 +15,128 @@
+  project         = var.project
+  count           = length(var.backend_params)
   name            = "${var.name}-backend-${count.index}"
-  port_name       = "${element(split(",", element(var.backend_params, count.index)), 1)}"
-  protocol        = "${var.backend_protocol}"
-  timeout_sec     = "${element(split(",", element(var.backend_params, count.index)), 3)}"
-  backend         = ["${var.backends["${count.index}"]}"]
-  health_checks   = ["${element(google_compute_http_health_check.default.*.self_link, count.index)}"]
-  security_policy = "${var.security_policy}"
-  enable_cdn      = "${var.cdn}"
+  port_name       = split(",", var.backend_params[count.index])[1]
+  protocol        = var.backend_protocol
+  timeout_sec     = split(",", var.backend_params[count.index])[3]
+  dynamic "backend" {
+    for_each = var.backends[count.index]
+    content {
+      balancing_mode               = lookup(backend.value, "balancing_mode", null)
+      capacity_scaler              = lookup(backend.value, "capacity_scaler", null)
+      description                  = lookup(backend.value, "description", null)
+      group                        = lookup(backend.value, "group", null)
+      max_connections              = lookup(backend.value, "max_connections", null)
+      max_connections_per_instance = lookup(backend.value, "max_connections_per_instance", null)
+      max_rate                     = lookup(backend.value, "max_rate", null)
+      max_rate_per_instance        = lookup(backend.value, "max_rate_per_instance", null)
+      max_utilization              = lookup(backend.value, "max_utilization", null)
+    }
+  }
+  health_checks   = [
+    google_compute_http_health_check.default[count.index].self_link]
+  security_policy = var.security_policy
+  enable_cdn      = var.cdn
 }
 
 resource "google_compute_http_health_check" "default" {
-  project      = "${var.project}"
-  count        = "${length(var.backend_params)}"
+  project      = var.project
+  count        = length(var.backend_params)
   name         = "${var.name}-backend-${count.index}"
-  request_path = "${element(split(",", element(var.backend_params, count.index)), 0)}"
-  port         = "${element(split(",", element(var.backend_params, count.index)), 2)}"
+  request_path = split(",", var.backend_params[count.index])[0]
+  port         = split(",", var.backend_params[count.index])[2]
 }
 
 # Create firewall rule for each backend in each network specified, uses mod behavior of element().
 resource "google_compute_firewall" "default-hc" {
-  count         = "${length(var.firewall_networks) * length(var.backend_params)}"
-  project       = "${element(var.firewall_projects, count.index) == "default" ? var.project : element(var.firewall_projects, count.index)}"
+  count         = length(var.firewall_networks) == 1 ? length(var.firewall_networks) * length(distinct(var.backend_params)) : length(var.firewall_networks) * length(var.backend_params)
```

Comment:
@aaron-lane , I agree it looks strange. Provisioning was failing in some examples and passing in others. I just included those case into the logic. 
The counter depends on number of firewall networks, since we migrated to dynamic allow block we don't need to create multiple instances of firewall objects counting networks and rules, instead we create multiple firewall rules inside the same FW resource. I just moved the old logic to a case when we have more than 1 firewall networks. But, actually, we only need to create 1 FW object per network.
I'll try to optimize it again with `count         = length(var.firewall_networks)`. I remember I tried, but I don't remember the reason why it didn't work.

Reply:
Did you have an opportunity to reconsider this logic? I'm still not clear on the structure or intent, like why a single firewall network uses a count based on the distinct set of backend parameters, but more firewall networks uses a count based on the entire list of backend parameters.